File: Colloc/14.04/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cctype>
#include <utility>

struct Token {
    bool is_number;
    int value;
    char op;    

    Token(int v) : is_number(true), value(v), op(0) {}
    Token(char c) : is_number(false), value(0), op(c) {}
};

std::vector<Token> tokenize(const std::string& s) {
    std::vector<Token> tokens;
    size_t i = 0, n = s.size();

    while (i < n) {
        char c = s[i];

        if (std::isspace(static_cast<unsigned char>(c))) {
            ++i;
        }
        else if (c == '+' || c == '-' || c == '(' || c == ')') {
            tokens.emplace_back(c);
            ++i;
        }
        else if (std::isdigit(static_cast<unsigned char>(c))) {
            int num = 0;
            while (i < n && std::isdigit(static_cast<unsigned char>(s[i]))) {
                num = num * 10 + (s[i] - '0');
                ++i;
            }
            tokens.emplace_back(num);
        }
        else {
            ++i;
        }
    }
    return tokens;
}

std::pair<int, size_t> eval_expr(const std::vector<Token>& tokens, size_t pos = 0) {
    int result = 0;
    int sign = 1;
    bool prev_was_value = false;
    size_t i = pos;

    while (i < tokens.size()) {
        const Token& tk = tokens[i++];
        
        if (tk.is_number) {
            result += sign * tk.value;
            prev_was_value = true;
        }
        else if (tk.op == '(') {
            auto [val, new_pos] = eval_expr(tokens, i);
            result += sign * val;
            prev_was_value = true;
            i = new_pos;
        }
        else if (tk.op == ')') {
            return { result, i };
        }
        else {
            if (tk.op == '+') {
                sign = 1;
            }
            else { 
                if (prev_was_value) {
                    sign = -1;       
                }
                else {
                    sign = -sign;       
                }
            }
            prev_was_value = false;
        }
    }

    return { result, i };
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string expr;
    std::getline(std::cin, expr);

    auto tokens = tokenize(expr);
    auto [result, _] = eval_expr(tokens, 0);

    std::cout << result << "\n";
    return 0;
}

File: Colloc/14.04/main.py
def tokenize(s):
    tokens = []
    i, n = 0, len(s)
    while i < n:
        c = s[i]
        if c.isspace():
            i += 1
        elif c in '+-()':
            tokens.append(c)
            i += 1
        elif c.isdigit():
            num = 0
            while i < n and s[i].isdigit():
                num = num * 10 + int(s[i])
                i += 1
            tokens.append(num)
    return tokens

def parse_term(tokens, i):
    if tokens[i] == '+':
        return parse_term(tokens, i+1)
    if tokens[i] == '-':
        val, j = parse_term(tokens, i+1)
        return -val, j
    if tokens[i] == '(':
        val, j = parse_expr(tokens, i+1)
        return val, j+1
    return tokens[i], i+1

def parse_expr(tokens, i=0):
    val, i = parse_term(tokens, i)
    while i < len(tokens) and tokens[i] in ('+', '-'):
        op = tokens[i]
        rhs, i = parse_term(tokens, i+1)
        if op == '+':
            val += rhs
        else:
            val -= rhs
    return val, i

expr = input().strip()
tokens = tokenize(expr)
result, pos = parse_expr(tokens)
print(result)
